packages/legend-query-builder/src/stores/watermark/QueryBuilderWatermarkStateBuilder.ts

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright (c) 2020-present, Goldman Sachs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  SimpleFunctionExpression,
+  matchFunctionName,
+} from '@finos/legend-graph';
+import { assertTrue, guaranteeType } from '@finos/legend-shared';
+import { QUERY_BUILDER_SUPPORTED_FUNCTIONS } from '../../graphManager/QueryBuilderSupportedFunctions.js';
+import type { QueryBuilderState } from '../QueryBuilderState.js';
+import { QueryBuilderValueSpecificationProcessor } from '../QueryBuilderStateBuilder.js';
+
+export const processWatermarkExpression = (
+  expression: SimpleFunctionExpression,
+  queryBuilderState: QueryBuilderState,
+): void => {
+  // check preceding expression
+  const precedingExpression = guaranteeType(
+    expression.parametersValues[0],
+    SimpleFunctionExpression,
+    `Can't process watermark() expression: only support watermark() immediately following an expression`,
+  );
+
+  assertTrue(
+    matchFunctionName(precedingExpression.functionName, [
+      QUERY_BUILDER_SUPPORTED_FUNCTIONS.GET_ALL,
+      QUERY_BUILDER_SUPPORTED_FUNCTIONS.FILTER,
+    ]),
+    `Can't process watermark() expression: only support watermark() immediately following either getAll() or filter()`,
+  );
+
+  QueryBuilderValueSpecificationProcessor.process(
+    precedingExpression,
+    queryBuilderState,
+  );
+
+  const watermarkLambda = expression.parametersValues[1];
```

Comment:
is it a `lambda`?  maybe `watermarkValue`

Reply:
true! 🙇